File: ios/Tiny/Sources/TopUp.swift
/**
 * 💧 TOP-UP PRESENTATION (iOS) — what to offer a user with no money, and what to
 * call the chain their money lives on.
 *
 * The third and last client of the self-hosted chain's copy work (web:
 * `lib/x402/top-up.ts`, Android: `WalletCore.kt`'s top-up block), ported
 * function-for-function so the three can't drift. iOS still renders
 * Coinbase + bridge.base.org on mainnet and faucet.circle.com on Sepolia — three
 * REAL-MONEY rails that are actively misleading on a chain we own. Nobody sells
 * TinyUSDC, and faucet.circle.com hands out SEPOLIA USDC, so a user who follows
 * any of them spends real money (or real time) and arrives holding a token this
 * deployment cannot credit. A stale on-ramp on our own chain isn't a dead link,
 * it's a trap. The in-house faucet (`POST /api/wallet/faucet`) is the only
 * source, and the server already says whether it exists
 * (`deposit_info.faucet.available`).
 *
 * PURE — no SwiftUI, no URLSession, no UIKit — because what's interesting here is
 * PRODUCT JUDGEMENT, not markup: which of three mutually-exclusive routes a
 * deployment offers, and how a refusal is phrased. Both look right in review and
 * are wrong in front of a user; you find out when someone is staring at a button
 * that 424s, or at "try again later" when the true answer is "you've had all of
 * it". So they get asserted (TopUpTests) instead of eyeballed.
 *
 * The one thing iOS has to decide for itself is PARSING. Every figure here
 * arrives as an `Any?` out of JSONSerialization, where a worker float is an
 * NSNumber and `Int(someDouble)` TRAPS on infinity/NaN/over-Int.max — a crash,
 * not a wrong number, on the money surface. See `micro(_:)`.
 */
import Foundation

enum TopUp {

    // MARK: - networks

    /// The three networks the payments stack settles on (worker `PayNetwork`).
    enum PayNetwork: String, CaseIterable {
        case base
        case baseSepolia = "base-sepolia"
        case tiny
    }

    /// Coerce an unknown network string, defaulting to the SAFEST reading (real
    /// Base): a client that guessed "trial" for an unknown name would label real,
    /// withdrawable money as un-withdrawable trial credit.
    static func asNetwork(_ raw: String?) -> PayNetwork {
        let n = (raw ?? "").trimmingCharacters(in: .whitespaces).lowercased()
        return PayNetwork(rawValue: n) ?? .base
    }

    /**
     * Label for a network, in the user's terms. `trial` is the load-bearing word:
     * both `tiny` and `base-sepolia` credit balance that is spendable inside tiny
     * but NOT withdrawable as real money, and a user who doesn't know that before
     * they earn on it will feel defrauded when the withdrawal is refused.
     */
    static func networkLabel(_ n: PayNetwork) -> String {
        switch n {
        case .tiny: return "Tiny Chain (trial credit)"
        case .baseSepolia: return "Base Sepolia (trial credit)"
        case .base: return "Base (real USDC)"
        }
    }

    /// Conservative display mapper for a quote/receipt `network` field: EXACT
    /// known names get the money-kind label ("Tiny Chain (trial credit)"),
    /// anything else renders verbatim — deliberately NOT through asNetwork(),
    /// whose default-to-base would label an unknown chain "real USDC", the one
    /// direction an approval card must never err.
    static func payNetworkDisplay(_ raw: String?) -> String? {
        guard let n = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !n.isEmpty else { return nil }
        guard let known = PayNetwork(rawValue: n) else { return n }
        return networkLabel(known)
    }

    /// Short form for a tight picker segment or a confirm dialog.
    static func networkShort(_ n: PayNetwork) -> String {
        switch n {
        case .tiny: return "Tiny Chain"
        case .baseSepolia: return "Sepolia"
        case .base: return "Base"
        }
    }

    /// True when balance earned on this network can leave as real money.
    static func isRealMoney(_ n: PayNetwork) -> Bool { n == .base }

    /**
     * Which networks a picker should show: only ever the deployment's OWN network
     * plus real Base, never all three.
     *
     * A deployment configures ONE chain, and offering the other trial network lets
     * a user paste a tx hash the receipt scanner can't see — the permanent "no
     * matching USDC transfer" 400 this repo already fixed once by seeding the
     * selector. Base stays listed on a trial deployment because a real deposit is
     * still the documented way to get withdrawable balance.
     *
     * iOS's pickers were a HARDCODED base/base-sepolia pair, so on a tiny-chain
     * deployment the user's own network wasn't offerable at all: every claim and
     * withdrawal was aimed at a chain their money isn't on.
     */
    static func networkChoices(_ defaultNetwork: String?) -> [PayNetwork] {
        let n = asNetwork(defaultNetwork)
        return n == .base ? [.base] : [n, .base]
    }

    // MARK: - money out of JSON

    /**
     * A micro-USDC figure from a JSON body, clamped so it can never trap.
     *
     * `Int(someDouble)` is a TRAP in Swift for infinity/NaN/over-`Int.max` — it
     * crashes rather than wrapping — and these numbers arrive from a worker as
     * JSON floats, where `1e999` decodes to `Double.infinity`. Same guard as
     * `Wallet.swift:207`, `loadPrice()` and `ChatStreamDecoder.safeMicro`;
     * deliberately a second copy rather than a shared call, because those two live
     * in targets this file isn't in (the watch app and the widget extension
     * compile ChatStreamDecoder, not the wallet) and a five-line trap-guard is
     * cheaper to duplicate than a cross-target dependency.
     *
     * Negative → 0: every field here is a non-negative allowance, and a negative
     * remaining that reached `min()` would enable a button the server refuses.
     */
    static func micro(_ value: Any?) -> Int {
        guard let n = value as? NSNumber else { return 0 }
        let d = n.doubleValue
        guard !d.isNaN else { return 0 }
        if d <= 0 { return 0 }
        return d >= Double(Int.max) ? Int.max : Int(d)
    }

    /**
     * micro-USDC → "$1.2" for a BUTTON. No min-2 padding, so a whole dollar reads
     * "$1" rather than "$1.00", and up to 6dp for the odd MIN-clamped drip. Web's
     * `usdShort` / Android's `priceLabel`.
     *
     * Locale-pinned: `String(format:)` uses the C locale for `%f`, which is what we
     * want here (a de_DE phone must not render the button as "$1,2" while the
     * server, the web app and Android all say "$1.2").
     */
    static func usdShort(_ micro: Int) -> String {
        var s = String(format: "%.6f", Double(micro) / 1_000_000)
        while s.hasSuffix("0") { s.removeLast() }
        if s.hasSuffix(".") { s.removeLast() }
        return "$\(s.isEmpty ? "0" : s)"
    }

    /// Human "2h 5m" until the next drip. Empty when it isn't in the future, so
    /// the caller can fall back to "after midnight UTC" rather than print "in 0m".
    static func untilNextDrip(_ seconds: Int) -> String {
        guard seconds > 0 else { return "" }
        let h = seconds / 3600
        let m = (seconds % 3600) / 60
        if h > 0 && m > 0 { return "\(h)h \(m)m" }
        if h > 0 { return "\(h)h" }
        // A sub-minute wait rounds UP: "in 0m" reads as a bug, and the drip really
        // is imminent.
        return "\(max(1, m))m"
    }

    // MARK: - the faucet block

    /// `deposit_info.faucet` (worker `PayDepositInfoCall`) — absent ⟺ no faucet.
    struct FaucetInfo: Equatable {
        var available = false
        var network: String?
        var dripMicro = 0
        var capMicro = 0
        var grantedMicro = 0
        var remainingMicro = 0
        var claimedToday = false
        var nextDripInSeconds = 0
        var reputation = 0
        var microPerPoint = 0
        var maxMicro = 0
    }

    /**
     * Parse the faucet block. A missing object is `nil` (no faucet); an
     * `{available:false}` parses to `available == false` rather than nil, so the
     * two server answers stay distinguishable — every consumer fails closed on
     * both, but "this deployment has no faucet" and "no card at all" are different
     * things to show.
     */
    static func parseFaucetInfo(_ raw: Any?) -> FaucetInfo? {
        guard let o = raw as? [String: Any] else { return nil }
        return FaucetInfo(
            // The worker sends a real JSON boolean, but a D1-ish 0/1 would read as
            // nil through `as? Bool` — and this one flag gates the whole card, so
            // read it through NSNumber's boolValue like the rest of the app does.
            available: truthy(o["available"]),
            network: o["network"] as? String,
            dripMicro: micro(o["drip_micro"]),
            capMicro: micro(o["cap_micro"]),
            grantedMicro: micro(o["granted_micro"]),
            // NOT defaulted to the cap: a missing `remaining_micro` must read as NO
            // credit (fail closed), or the button enables and 400s on press.
            remainingMicro: micro(o["remaining_micro"]),
            claimedToday: truthy(o["claimed_today"]),
            nextDripInSeconds: micro(o["next_drip_in_seconds"]),
            reputation: micro(o["reputation"]),
            microPerPoint: micro(o["micro_per_point"]),
            maxMicro: micro(o["max_micro"]))
    }

    /// A JSON flag that may arrive as `true` or as a D1 integer 1.
    static func truthy(_ value: Any?) -> Bool {
        if let b = value as? Bool { return b }
        if let n = value as? NSNumber { return n.boolValue }
        return false
    }

    // MARK: - the route

    /**
     * Which top-up route this deployment offers. Exactly one, always — mutually
     * exclusive ON PURPOSE. A fiat card button offered "just in case" beside the
     * faucet, on a chain where the card cannot deliver, IS the bug this deletes.
     *
     *  - `faucet`  — we own the chain, so we issue the credit.
     *  - `testnet` — Sepolia: the public faucet is the one true source; fiat
     *                on-ramps deliver MAINNET USDC the claim scanner can't see.
     *  - `fiat`    — real Base: cards/bridges work and a faucet would be nonsense.
     */
    enum Route { case faucet, testnet, fiat }

    /**
     * Pick the route.
     *
     * Keyed on `faucet.available` — the server's own answer — and NOT on
     * `chain == "tiny"`, because those two legitimately disagree: the faucet needs
     * a mintable token AND a deployer key, so a half-configured tiny-chain
     * deployment reports `tiny` with no faucet, and a claim button there 424s on
     * every press. Fall back to what the network can actually do.
     */
    static func route(chain: String?, faucet: FaucetInfo?) -> Route {
        if faucet?.available == true { return .faucet }
        return asNetwork(chain) == .baseSepolia ? .testnet : .fiat
    }

    /// The top-up card's blurb, one per route (web + Android parity).
    static func blurb(_ route: Route) -> String {
        switch route {
        case .faucet:
            return "This deployment runs its own chain, so credit comes straight from us — no card, no exchange, no wallet needed. It's spendable on any tiny; it isn't withdrawable as real USDC."
        case .testnet:
            return "This deployment runs on Base Sepolia — grab free testnet USDC, then claim it below to try risk-free. (Testnet credits aren't withdrawable as real USDC.)"
        case .fiat:
            return "Buy or bridge USDC on Base into your linked address, then claim it below."
        }
    }

    /// The card's own title — the faucet route isn't "Get USDC", it's credit.
    static func title(_ route: Route) -> String {
        route == .faucet ? "💧 Get credit (free daily top-up)" : "💳 Get USDC"
    }

    /**
     * The EXTERNAL deposit sources, reached only on the TESTNET and FIAT routes.
     * On the FAUCET route this is deliberately EMPTY — the card renders a claim
     * button instead, and no external rail can deliver a token only this
     * deployment mints.
     */
    static func usdcSources(_ route: Route) -> [(label: String, url: String)] {
        switch route {
        case .faucet: return []
        case .testnet: return [("🧪 Get free testnet USDC", "https://faucet.circle.com")]
        case .fiat: return [("Coinbase", "https://www.coinbase.com/price/usdc"),
                            ("Bridge to Base", "https://bridge.base.org")]
        }
    }

    // MARK: - the claim button

    struct FaucetCta: Equatable {
        var enabled: Bool
        var label: String
        /// Why it's disabled — "" when enabled.
        var reason: String
    }

    /**
     * Three states, and the two REFUSALS must never collapse into one sentence.
     * They're the client mirror of the worker's deliberately-distinct 429
     * (`already_claimed`) vs 400 (`ceiling_reached`): "wait until UTC midnight"
     * and "you've spent your lifetime ceiling, get followed to raise it" are
     * OPPOSITE instructions, so a shared "try again later" sends a permanently
     * capped user back to the button every day, and a user who merely claimed
     * today off to farm reputation they don't need.
     *
     * Ceiling is checked BEFORE the daily claim: someone fully capped AND claimed
     * today is capped — "come back tomorrow" would be a lie, because tomorrow's
     * drip is refused too.
     */
    static func faucetCta(_ f: FaucetInfo?) -> FaucetCta {
        guard let f, f.available else {
            return FaucetCta(enabled: false, label: "Top-up unavailable",
                             reason: "This deployment has no in-house faucet.")
        }
        let remaining = max(0, f.remainingMicro)
        if remaining <= 0 {
            return FaucetCta(
                enabled: false, label: "Lifetime credit used",
                // The actionable half: the ceiling is reputation-scaled, so it GROWS.
                reason: "You've used all \(usdShort(f.capMicro)) of your trial credit. Get followed to raise the ceiling, or deposit real USDC on Base.")
        }
        if f.claimedToday {
            let wait = untilNextDrip(f.nextDripInSeconds)
            return FaucetCta(
                enabled: false, label: "Claimed today",
                reason: "Next top-up \(wait.isEmpty ? "after midnight UTC" : "in \(wait)") — \(usdShort(remaining)) still left on your ceiling.")
        }
        // The worker credits MIN(drip, remaining), so the button must promise what
        // will actually land — "Claim $1" that pays $0.30 is the client breaking a
        // promise the server never made.
        let credit = min(max(0, f.dripMicro), remaining)
        return FaucetCta(enabled: true, label: "Claim \(usdShort(credit)) free credit", reason: "")
    }

    /**
     * The one-line explanation of the ceiling. Separate from the CTA because it's
     * shown in ALL faucet states — a user who just claimed still needs to know why
     * their ceiling is what it is, and that being followed is what raises it.
     */
    static func ceilingNote(_ f: FaucetInfo?) -> String {
        guard let f, f.available else { return "" }
        let per = usdShort(f.microPerPoint)
        let rep = f.reputation
        let earned = rep > 0
            ? " Your \(rep) reputation \(rep == 1 ? "point adds" : "points add") \(per) each"
            : " Earn reputation (\(per) per point) by getting followed"
        return "\(usdShort(f.grantedMicro)) of \(usdShort(f.capMicro)) used.\(earned), up to \(usdShort(f.maxMicro))."
    }

    // MARK: - the claim reply

    /**
     * The outcome of `POST /api/wallet/faucet`.
     *
     * The two refusals stay SEPARATE cases rather than one failure with a flag,
     * for the same reason `faucetCta` keeps two messages — and the route passes
     * the worker's own sentences through verbatim, so re-wording them here would
     * undo that on purpose.
     */
    enum FaucetResult: Equatable {
        /// Credited. `reserveBacked` = the matching TinyUSDC mint landed on-chain
        /// (best-effort by design — the credit is real either way).
        case ok(creditedMicro: Int, reserveBacked: Bool, explorer: String?)
        /// 429 — today's drip is spent; the wording carries `next_drip_in_seconds`.
        case alreadyClaimed(String)
        /// 400 — the reputation-scaled lifetime ceiling is spent; waiting won't help.
        case ceilingReached(String)
        /// 424 (no faucet here) / 401 / 500 — everything else.
        case failed(String)
    }

    /// `status` is the HTTP code when the caller could capture it (Wallet's
    /// `post()` injects `_status`), used only as a fallback: the worker sets the
    /// flags explicitly, and a proxy that rewrote the body is the case the code is
    /// for.
    static func parseFaucetResult(_ body: [String: Any]?) -> FaucetResult {
        guard let d = body else { return .failed("couldn't reach the faucet") }
        if truthy(d["ok"]) {
            return .ok(creditedMicro: micro(d["credited_micro"]),
                       reserveBacked: truthy(d["reserve_backed"]),
                       explorer: d["explorer"] as? String)
        }
        let err = (d["error"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "faucet unavailable"
        // Flag first, status second — and ceiling BEFORE already-claimed, matching
        // faucetCta: a capped user who also claimed today must hear about the cap.
        if truthy(d["ceiling_reached"]) { return .ceilingReached(err) }
        if truthy(d["already_claimed"]) { return .alreadyClaimed(err) }
        if micro(d["_status"]) == 429 { return .alreadyClaimed(err) }
        return .failed(err)
    }
}

File: web/lib/x402/top-up.ts
/**
 * 💧 TOP-UP PRESENTATION — how a client decides what to offer a user who has no
 * money, and what to say about the network their money lives on.
 *
 * This is the client half of the self-hosted chain (docs/e2e-gaps-report-
 * 2026-07-25.md §1.2 item 7). Every client still renders Coinbase Onramp,
 * MoonPay and faucet.circle.com — three real-money on-ramps that are **actively
 * misleading** on a chain we own: nobody sells TinyUSDC, and faucet.circle.com
 * hands out Sepolia USDC, so a user who follows any of those links spends real
 * money (or real time) and arrives with a token this deployment cannot credit.
 * The in-house faucet (`/api/wallet/faucet`) is the only source, and the server
 * already advertises whether it exists via `deposit_info.faucet.available`.
 *
 * PURE (env-free, DOM-free, no fetch) because the interesting content here is
 * PRODUCT JUDGEMENT, not markup: which of three mutually-exclusive top-up routes
 * a deployment offers, and how a refusal is phrased. Both are the kind of thing
 * that looks right in review and is wrong in front of a user — you find out when
 * someone is staring at a button that 424s, or at "try again later" when the real
 * answer is "you've had all of it". So they get asserted (tests/top-up.test.ts)
 * rather than eyeballed, and all three clients can converge on one source.
 *
 * The three routes are mutually exclusive ON PURPOSE. Offering a fiat card
 * button "just in case" next to the faucet, on a chain where the card can't
 * deliver, is the exact bug this module exists to delete.
 */

/** Networks the payments stack can settle on — mirrors the worker's `PayNetwork`. */
export type PayNetwork = 'base' | 'base-sepolia' | 'tiny'

/** What `deposit_info.faucet` looks like (worker: PayDepositInfoCall). */
export type FaucetInfo = {
  available?: boolean
  network?: string
  drip_micro?: number
  cap_micro?: number
  granted_micro?: number
  remaining_micro?: number
  claimed_today?: boolean
  next_drip_in_seconds?: number
  reputation?: number
  micro_per_point?: number
  max_micro?: number
}

export type DepositInfoLike = {
  default_network?: string
  linked_address?: string | null
  faucet?: FaucetInfo | null
}

/**
 * Which top-up route this deployment offers. Exactly one, always:
 *
 *  - `faucet`  — we own the chain, so we issue the credit. No external rail can
 *                deliver a token only we can mint.
 *  - `testnet` — Sepolia: the public faucet is the one true source and fiat
 *                on-ramps deliver MAINNET USDC the claim flow will reject.
 *  - `fiat`    — real Base: cards/bridges work and a faucet would be nonsense.
 */
export type TopUpRoute = 'faucet' | 'testnet' | 'fiat'

/** Coerce an unknown network string, defaulting to the safest reading. */
export const asNetwork = (raw: unknown): PayNetwork => {
  const n = String(raw ?? '').toLowerCase().trim()
  return n === 'tiny' || n === 'base-sepolia' ? n : 'base'
}

/**
 * Pick the route.
 *
 * Keyed on `faucet.available` — the server's own answer — and NOT on
 * `default_network === 'tiny'`, because those two can legitimately disagree: the
 * faucet needs a mintable token AND a deployer key, so a half-configured
 * tiny-chain deployment reports `tiny` with no faucet. Trusting the network
 * string there would render a claim button that 424s every time, which is the
 * failure this whole item is about. Fall back to what the network can actually
 * do instead.
 */
export const topUpRoute = (info: DepositInfoLike | null | undefined): TopUpRoute => {
  if (info?.faucet?.available) return 'faucet'
  return asNetwork(info?.default_network) === 'base-sepolia' ? 'testnet' : 'fiat'
}

/** Micro-USDC → "$1.20". Trailing zeros trimmed past cents, junk → "$0". */
export const usdShort = (micro: unknown): string => {
  const n = Number(micro)
  const v = Number.isFinite(n) ? n / 1_000_000 : 0
  // Cents for whole/2dp amounts, up to 6dp for the odd clamped drip, and never
  // "$1.00" where "$1" reads better on a button.
  const s = v.toFixed(6).replace(/0+$/, '').replace(/\.$/, '')
  return `$${s}`
}

/** Human "2h 5m" until the next drip. Empty when it isn't in the future. */
export const untilNextDrip = (seconds: unknown): string => {
  const n = Math.floor(Number(seconds))
  if (!Number.isFinite(n) || n <= 0) return ''
  const h = Math.floor(n / 3600)
  const m = Math.floor((n % 3600) / 60)
  if (h && m) return `${h}h ${m}m`
  if (h) return `${h}h`
  return `${Math.max(1, m)}m`
}

/**
 * What the faucet button says and whether it's live.
 *
 * Three states, and the two refusals must NEVER collapse into one sentence —
 * they're the client mirror of the worker's deliberately-distinct 429 vs 400
 * (see PayFaucetCall). "Wait until UTC midnight" and "you've spent your lifetime
 * ceiling, earn more by getting followed" are opposite instructions; a shared
 * "try again later" sends a user who is permanently capped back to the button
 * every day, and a user who just claimed off to farm reputation they don't need.
 *
 * Ceiling is checked BEFORE the daily claim: someone who is fully capped AND
 * claimed today is capped — telling them to come back tomorrow would be a lie,
 * because tomorrow's drip is refused too.
 *
 * ⚠️ `claimed_today` IS A SERVER OBSERVATION WITH AN EXPIRY, and this function
 *    used to treat it as a standing fact. It means "already claimed for the UTC
 *    day that was current when deposit_info was fetched" — so once that day
 *    rolls over it is simply out of date, and a page left open across UTC
 *    midnight kept the Claim button disabled forever with no visible reason.
 *    The caller may now pass `remainingSeconds` (from
 *    faucet-countdown.dripRemainingSeconds, which is derived from a LIVE clock
 *    rather than the frozen server delta). At 0 the claim is offered again; the
 *    faucet's idempotency key is per-UTC-day, so the server accepts it, and if
 *    the boundary is somehow missed the 429 is already surfaced verbatim.
 *    Omitting the option preserves the old behaviour exactly, for callers that
 *    have no clock — the number of them is the point of the option.
 */
export type FaucetCta = {
  enabled: boolean
  label: string
  /** Why it's disabled — '' when enabled. */
  reason: string
}

export const faucetCta = (
  f: FaucetInfo | null | undefined,
  opts?: { remainingSeconds?: number | null },
): FaucetCta => {
  if (!f?.available) {
    return { enabled: false, label: 'Top-up unavailable', reason: 'This deployment has no in-house faucet.' }
  }
  const remaining = Math.max(0, Number(f.remaining_micro) || 0)
  if (remaining <= 0) {
    return {
      enabled: false,
      label: 'Lifetime credit used',
      // The actionable half: the ceiling is reputation-scaled, so it GROWS.
      reason: `You've used all ${usdShort(f.cap_micro)} of your trial credit. Get followed to raise the ceiling, or deposit real USDC on Base.`,
    }
  }
  // `undefined` = the caller has no clock, so fall back to the server's own
  // frozen delta and behave exactly as before. `null` = a clock-aware caller
  // that has no deadline to count to (the server sent no delta), which is still
  // "wait", just without a duration to name.
  const live = opts?.remainingSeconds
  const lapsed = live != null && Number.isFinite(live) && live <= 0
  if (f.claimed_today && !lapsed) {
    const wait = untilNextDrip(live === undefined ? f.next_drip_in_seconds : live)
    return {
      enabled: false,
      label: 'Claimed today',
      reason: `Next top-up ${wait ? `in ${wait}` : 'after midnight UTC'} — ${usdShort(remaining)} still left on your ceiling.`,
    }
  }
  // The drip can be MIN-clamped by the remaining ceiling, so the button must
  // promise what will actually be credited, not the nominal drip.
  const credit = Math.min(Number(f.drip_micro) || 0, remaining)
  return { enabled: true, label: `Claim ${usdShort(credit)} free credit`, reason: '' }
}

/**
 * 💧 What the faucet card says when the claim came back with NOTHING IT COULD READ.
 *
 * All three clients used to say **"couldn't reach the faucet"** here, under a ⚠️,
 * and two of them added "try again". None of them can back any of that up:
 *
 *   - web  — `faucetClaim()` is `fetch().then(r => r.json())`, so the rejection is
 *            a transport drop OR the `AbortSignal.timeout` firing (the request WAS
 *            delivered) OR a body that wasn't JSON (the server DID answer — a
 *            platform 502 page, a captive portal's login HTML).
 *   - iOS  — `Wallet.post()` returns nil for a malformed `Api.base` (Settings typo,
 *            nothing was ever sent), for transport/timeout, and for a non-JSON
 *            body. `LoadFailure.message` already knows to keep those apart.
 *   - Android — `executeJson` never throws on non-JSON, so its null is transport
 *            OR the settle timeout. Still not "unreachable".
 *
 * ⚠️ And the timeout case is the one that costs money. `/api/wallet/faucet`
 *    CREDITS THE LEDGER and only then waits on the TinyUSDC mint receipt (~20s),
 *    so a client that gives up is looking at credit the user already has. Told
 *    "couldn't reach the faucet — try again", they press again, get the 429, and
 *    now believe the claim failed twice while holding the money. iOS's
 *    `claimFaucet` documents that exact sequence — it raised its own deadline to
 *    120s to make it rarer and left the sentence in place.
 *
 * So: name no cause, don't invite the retry that 429s, and point at the balance,
 * which is the only thing that actually answers the question. Same shape the
 * withdraw path on all three clients already uses for its own unknown outcome
 * ("couldn't confirm — check Activity before retrying") — the faucet, one card up,
 * never got it. The ⏳ is load-bearing: ⚠️ is the unfounded conclusion in glyph
 * form, and the picture is read before the words.
 */
export const faucetNoAnswerNote =
  "⏳ Couldn't confirm the claim — the drip may already be credited. Check your balance before claiming again."

/**
 * The one-line explanation of the ceiling. Named separately from the CTA because
 * it's shown in ALL faucet states — a user who just claimed still needs to know
 * why their ceiling is what it is, and that following is what raises it.
 */
export const ceilingNote = (f: FaucetInfo | null | undefined): string => {
  if (!f?.available) return ''
  const cap = usdShort(f.cap_micro)
  const used = usdShort(f.granted_micro)
  const rep = Number(f.reputation) || 0
  const per = usdShort(f.micro_per_point)
  const max = usdShort(f.max_micro)
  const earned = rep > 0 ? ` Your ${rep} reputation ${rep === 1 ? 'point adds' : 'points add'} ${per} each` : ` Earn reputation (${per} per point) by getting followed`
  return `${used} of ${cap} used.${earned}, up to ${max}.`
}

/**
 * Label for a network, in the user's terms. `trial` is the load-bearing word:
 * both `tiny` and `base-sepolia` credit balance that is spendable inside tiny
 * but NOT withdrawable as real money, and a user who doesn't know that before
 * they earn on it will feel defrauded when the withdrawal is refused.
 */
export const networkLabel = (n: PayNetwork): string =>
  n === 'tiny' ? 'Tiny Chain (trial credit)'
    : n === 'base-sepolia' ? 'Base Sepolia (trial credit)'
      : 'Base (real USDC)'

/** Short form for a tight picker chip. */
export const networkShort = (n: PayNetwork): string =>
  n === 'tiny' ? 'Tiny Chain' : n === 'base-sepolia' ? 'Sepolia' : 'Base'

/** True when balance earned on this network can leave as real money. */
export const isRealMoney = (n: PayNetwork): boolean => n === 'base'

/**
 * The kind of money a payment on this network spends, for any sentence that
 * asks a user to approve one. "trial credit" vs "real USDC" is the difference
 * between a sandbox tap and a spend — a confirm prompt that omits it makes the
 * user approve blind (the same rule networkLabel enforces on the pickers).
 */
export const moneyKind = (n: PayNetwork): string =>
  isRealMoney(n) ? 'real USDC' : 'trial credit'

/** The one sentence a quote card / agent echoes while awaiting approval. */
export const quoteSummary = (amountMicro: number, n: PayNetwork): string =>
  `Pay ${usdShort(amountMicro)} in ${moneyKind(n)} to consult this service over x402. Awaiting your approval.`

/**
 * Suffix for the 402 challenge's human-readable description. Three networks,
 * three answers — the old `net === 'base' ? '' : ' (testnet)'` ternary branded
 * a self-hosted chain "(testnet)", which tells an external payer to expect a
 * public faucet that does not exist (the c26 two-branch-ternary bug, on copy).
 */
export const x402DescSuffix = (n: PayNetwork): string =>
  n === 'base' ? '' : n === 'base-sepolia' ? ' (testnet)' : ' (trial credit)'

/**
 * Which networks a picker should show.
 *
 * Only ever the deployment's own network plus real Base — never all three. A
 * deployment configures ONE chain; offering the other trial network would let a
 * user paste a tx hash the receipt scanner can't see (the permanent "no matching
 * USDC transfer" 400 this repo already fixed once by seeding the selector).
 * Base stays listed on a trial deployment because a real deposit is still the
 * documented way to get withdrawable balance.
 */
export const networkChoices = (defaultNetwork: unknown): PayNetwork[] => {
  const n = asNetwork(defaultNetwork)
  return n === 'base' ? ['base'] : [n, 'base']
}

/**
 * 📖 THE WALLET EXPLAINER, per network.
 *
 * The "What is the tiny wallet?" card is the longest piece of money copy we
 * ship and, until now, the only one still written as if every deployment were
 * real Base: "Deposits and withdrawals are real USDC on Base", "Buy or send
 * USDC on Base", "other people's agents can discover your tiny and pay it".
 * On a chain we host, all three are the same lie c-g deleted from the on-ramps
 * and c27 deleted from the agent's own mouth — and this one is worse than a
 * dead link, because a user reads it BEFORE they earn, decides the balance is
 * real money, and finds out otherwise at the withdrawal.
 *
 * Two of the three sentences are network-dependent for DIFFERENT reasons, so
 * they're separate fields rather than one paragraph:
 *
 *  - `custody`  — can this balance leave as real money? (`isRealMoney`)
 *  - `reach`    — can an outside agent pay it? x402 is open on every network,
 *                 but an external payer needs the ASSET, and on our chain the
 *                 only mint is ours. So the honest claim shrinks from "the whole
 *                 internet" to "any agent we've issued credit to" (report item
 *                 10, which we already corrected in the agent's prose).
 *  - `quickStart` — the deposit instruction; a faucet deployment has no "send
 *                 USDC" step at all, so telling them to buy some is a dead end.
 *
 * Keyed on the NETWORK (not faucet availability) because these are claims about
 * what the money IS, which stays true whether or not the faucet is up — the
 * opposite of `topUpRoute`, which is keyed on capability because it renders a
 * button that must actually work.
 */
export type WalletIntro = {
  custody: string
  reach: string
  quickStart: string
  /** Heading for the custody bullet — "Real money, your custody" is only true on Base. */
  custodyTitle: string
}

export const walletIntro = (network: unknown): WalletIntro => {
  const n = asNetwork(network)
  if (n === 'base') {
    return {
      custodyTitle: 'Real money, your custody',
      custody:
        'Deposits and withdrawals are real USDC on Base (an Ethereum L2 by Coinbase). Withdrawals are instant and self-serve — funds go only to the wallet address YOU linked, so a stolen session can’t redirect your money.',
      reach:
        'Priced tinys are also payable by ANY AI agent via the open x402 protocol — other people’s agents can discover your tiny and pay it per request in USDC, no tiny.technology account needed. Your AI becomes an API that earns.',
      quickStart:
        '1) Link your wallet address below · 2) Buy or send USDC on Base · 3) Claim the deposit · 4) Chat with paid tinys, or price your own and start earning.',
    }
  }
  const chain = networkShort(n)
  return {
    custodyTitle: 'Trial credit, not real money',
    custody: `Balance here is ${moneyKind(n)} on ${chain} — it spends on any paid tiny, but it cannot be withdrawn as real USDC, and earning it doesn’t make it withdrawable. Deposit real USDC on Base if you want a balance you can cash out.`,
    reach: `Priced tinys are payable by any AI agent over the open x402 protocol — but a payer needs ${chain} credit to pay with, and on this chain we’re the only source of it. So your audience is agents we’ve issued credit to, not the open internet.`,
    quickStart:
      '1) Claim your free daily credit below · 2) Chat with paid tinys, or price your own and start earning · 3) Watch the settlements land on the chain explorer.',
  }
}
